src/util/kafka/utility.js

```diff
@@ -0,0 +1,429 @@
+ * @param {string} metadataAction - the action flow. Example: 'prepare'
+ * @param {object} state - the state of the message being passed.
+ * Example:
+ * SUCCESS: {
+ *   status: 'success',
+ *   code: 0,
+ *   description: 'action successful'
+ * }
+ *
+ * @returns {object} - Returns updated messageProtocol
+ */
+const updateMessageProtocolMetadata = (messageProtocol, metadataType, metadataAction, state) => {
+  if (!messageProtocol.metadata) {
+    messageProtocol.metadata = {
+      event: {
+        id: Uuid(),
+        type: metadataType,
+        action: metadataAction,
+        state: state
+      }
+    }
+  } else {
+    messageProtocol.metadata.event.responseTo = messageProtocol.metadata.event.id
+    messageProtocol.metadata.event.id = Uuid()
+    messageProtocol.metadata.event.type = metadataType
+    messageProtocol.metadata.event.action = metadataAction
+    messageProtocol.metadata.event.state = state
+  }
+  return messageProtocol
+}
+
+/**
+ * @function createStreamingMessage
+ *
+ * @param {string} id - the unique ID for the correlation of messages
+ * @param {string} to - to whom the message is going
+ * @param {string} from - from whom the message is received
+ * @param {object} metadata - The metadata for streaming
+ * @param {object} payload - The payload of the message
+ * @param {object} headers - headers from the request
+ * @param {object} uriParams - the URI parameters passed in request.
+ *
+ * @returns {object} - Returns generated messageProtocol
+ */
+const createStreamingMessage = (id, to, from, metadata, headers, payload, uriParams = undefined) => {
+  return {
+    id,
+    to,
+    from,
+    type: Enum.Http.Headers.DEFAULT.APPLICATION_JSON,
+    content: {
+      uriParams: uriParams || undefined,
+      headers: headers,
+      payload: payload || {}
+    },
+    metadata
+  }
+}
+
+/**
+ * @function createStreamingMessageFromRequest
+ * @param {string} id - the unique ID for the correlation of messages
+ * @param {object} request - The current messageProtocol from kafka
+ * @param {string} to - the action flow. Example: 'prepare'
+ * @param {string} from - the state of the message being passed.
+ * @param {object} metadata - The metadata for streaming
+ *
+ * @returns {object} - Returns generated messageProtocol
+ */
+const createStreamingMessageFromRequest = (id, request, to, from, metadata) => {
+  return createStreamingMessage(id, to, from, metadata, request.headers, request.dataUri, request.params)
+}
+
+/**
+ * @function createMetadata
+ *
+ * @param {object} correlationId - The current messageProtocol from kafka
+ * @param {string} type - the type of message being sent Example: 'fulfil'
+ * @param {string} action - the action of message being sent Example: 'commit'
+ * @param {object} state - The state object for metadata usually contains information of status
+ *
+ * @returns {object} - Returns generated metadata
+ */
+const createMetadata = (correlationId, type, action, state) => {
```

Comment:
this should be 

1. `const createMetadata(correlationId, type, action, state) ` ==> `const createMetadata = (correlationId, eventMetadata) => {`

2. Then have a method `createEventMetadata(type, action, eventState) `.

3. Then have a method `createEventState(status, code, description) `.

4. add another method `createMetadataWithCorrelatedEvent(correlationId, type, action, state) 

5. Ensure 4 supports overriding shortcut so that we can also provide the eventState directly -> `createMetadataWithCorrelatedEvent(correlationId, type, action, status, code, description). Aka, check if "state" is an object, otherwise check for the status, code & description.